### PR TITLE
[BOOT] Support compiling freeldr as a UEFI loader as well - Implement UI functions

### PR DESCRIPTION
**Pull request by DarkFire01**
## Purpose

First batch of changes to implement a UEFI version of freeldr

## Proposed changes

- Compile freeldr as EFI binary on top of the existing loader
- Stub out various functions so we can create a UEFI mach type in freeldr
- Implement all of the video output function so we can display a pretty freeldr BSoD

### CODE REVIEW

File: boot/freeldr/uefi/uefivid.c [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+VOID
+UefiVideoClearScreen(UCHAR Attr)
+{
+
+    ULONG FgColor, BgColor;
+
+    UefiVideoAttrToColors(Attr, &FgColor, &BgColor);
+    for(int y = 0; y < framebufferData.ScreenHeight; y++)
+    {
+        for(int x = 0; x < framebufferData.ScreenWidth; x++)
```
Comment (HBelusca) [inline boot/freeldr/uefi/uefivid.c:78]:
```suggestion
        for (x = 0; x < framebufferData.ScreenWidth; ++x)
```

File: boot/freeldr/uefi/uefivid.c [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+UefiPrintFramebufferData()
+{
+    TRACE("Framebuffer BaseAddress       : %X\n", framebufferData.BaseAddress);
+    TRACE("Framebuffer BufferSize        : %X\n", framebufferData.BufferSize);
+    TRACE("Framebuffer ScreenWidth       : %d\n", framebufferData.ScreenWidth);
+    TRACE("Framebuffer ScreenHeight      : %d\n", framebufferData.ScreenHeight);
+    TRACE("Framebuffer PixelsPerScanLine : %d\n", framebufferData.PixelsPerScanLine);
+    TRACE("Framebuffer PixelFormat       : %d\n", framebufferData.PixelFormat);
+}
+ULONG
```
Comment (HBelusca) [inline boot/freeldr/uefi/uefivid.c:50]:
```suggestion

static ULONG
```
static if that function is only used within this file.

File: boot/freeldr/uefi/uefivid.c [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+    UCHAR Intensity;
+    Intensity = (0 == (Attr & 0x08) ? 127 : 255);
+
+    return 0xff000000 |
+         (0 == (Attr & 0x04) ? 0 : (Intensity << 16)) |
+         (0 == (Attr & 0x02) ? 0 : (Intensity << 8)) |
+         (0 == (Attr & 0x01) ? 0 : Intensity);
+}
+
+VOID
```
Comment (HBelusca) [inline boot/freeldr/uefi/uefivid.c:62]:
```suggestion
static VOID
```

File: boot/freeldr/uefi/uefivid.c [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    *Depth =  0;
+}
+
+VIDEODISPLAYMODE
+UefiVideoSetDisplayMode(char *DisplayMode, BOOLEAN Init)
+{
+    /* We only have one mode, semi-text */
+    return VideoTextMode;
+}
+ULONG
```
Comment (HBelusca) [inline boot/freeldr/uefi/uefivid.c:139]:
```suggestion

ULONG
```

File: boot/freeldr/uefi/uefivid.c [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    return ((framebufferData.ScreenHeight - 2) / CHAR_HEIGHT * (framebufferData.ScreenWidth / CHAR_WIDTH) * 2);
+}
+
+VOID
+UefiVideoCopyOffScreenBufferToVRAM(PVOID Buffer)
+{
+    PUCHAR OffScreenBuffer = (PUCHAR) Buffer;
+
+    ULONG Col, Line;
+    for (Line = 0; Line < (framebufferData.ScreenHeight - 2) / CHAR_HEIGHT; Line++)
```
Comment (HBelusca) [inline boot/freeldr/uefi/uefivid.c:151]: Question (also valid for other places): why do we need to subtract two?

File: boot/freeldr/uefi/uefivid.c [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+UefiVideoIsPaletteFixed(VOID)
+{
+    return 0;
+}
+
+VOID
+UefiVideoSetPaletteColor(UCHAR Color, UCHAR Red,
+                         UCHAR Green, UCHAR Blue)
+{
+      /* Not supported */
```
Comment (HBelusca) [inline boot/freeldr/uefi/uefivid.c:204]:
```suggestion
    /* Not supported */
```

File: boot/freeldr/uefi/uefivid.c [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+}
+
+VOID
+UefiVideoGetPaletteColor(UCHAR Color, UCHAR* Red,
+                         UCHAR* Green, UCHAR* Blue)
+{
+      /* Not supported */
```
Comment (HBelusca) [inline boot/freeldr/uefi/uefivid.c:211]:
```suggestion
    /* Not supported */
```

File: boot/freeldr/uefi/uefivid.c [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+ULONG
+UefiVideoGetBufferSize(VOID)
+{
+    return ((framebufferData.ScreenHeight - 2) / CHAR_HEIGHT * (framebufferData.ScreenWidth / CHAR_WIDTH) * 2);
+}
+
+VOID
+UefiVideoCopyOffScreenBufferToVRAM(PVOID Buffer)
+{
+    PUCHAR OffScreenBuffer = (PUCHAR) Buffer;
```
Comment (HBelusca) [inline boot/freeldr/uefi/uefivid.c:148]:
```suggestion
    PUCHAR OffScreenBuffer = (PUCHAR)Buffer;
```

File: boot/freeldr/uefi/uefivid.c [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+ULONG
+UefiVideoAttrToSingleColor(UCHAR Attr)
+{
+    UCHAR Intensity;
+    Intensity = (0 == (Attr & 0x08) ? 127 : 255);
+
+    return 0xff000000 |
+         (0 == (Attr & 0x04) ? 0 : (Intensity << 16)) |
+         (0 == (Attr & 0x02) ? 0 : (Intensity << 8)) |
+         (0 == (Attr & 0x01) ? 0 : Intensity);
```
Comment (HBelusca) [inline boot/freeldr/uefi/uefivid.c:59]:
```suggestion
           (0 == (Attr & 0x04) ? 0 : (Intensity << 16)) |
           (0 == (Attr & 0x02) ? 0 : (Intensity << 8)) |
           (0 == (Attr & 0x01) ? 0 : Intensity);
```

File: boot/freeldr/uefi/uefivid.c [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+#include <debug.h>
+DBG_DEFAULT_CHANNEL(WARNING);
+
+#define CHAR_WIDTH  8
+#define CHAR_HEIGHT 16
+#define TOP_BOTTOM_LINES 0
+
+/* GLOBALS ********************************************************************/
+
+extern EFI_SYSTEM_TABLE * GlobalSystemTable;
```
Comment (HBelusca) [inline boot/freeldr/uefi/uefivid.c:19]:
```suggestion
extern EFI_SYSTEM_TABLE* GlobalSystemTable;
```

File: boot/freeldr/uefi/uefisetup.c [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+#include <uefildr.h>
+
+#include <debug.h>
+DBG_DEFAULT_CHANNEL(WARNING);
+
+/* GLOBALS ********************************************************************/
+
+EFI_GUID EfiGraphicsOutputProtocol = EFI_GRAPHICS_OUTPUT_PROTOCOL_GUID;
+extern EFI_SYSTEM_TABLE *GlobalSystemTable;
```
Comment (HBelusca) [inline boot/freeldr/uefi/uefisetup.c:16]:
```suggestion
extern EFI_SYSTEM_TABLE* GlobalSystemTable;
```

File: boot/freeldr/uefi/uefisetup.c [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+{
+    EFI_STATUS Status = 0;
+    Status = GlobalSystemTable->BootServices->AllocatePool(EfiLoaderData, sizeof(MachVtbl), (void**)&MachVtbl);
+    if (UefiHandleService(Status) != EFI_SUCCESS)
+        ERR("UefiMachInit: Could not allocated memory for MachVtbl");
+
+    /* Setup GOP */
+    Status = GlobalSystemTable->BootServices->LocateProtocol(&EfiGraphicsOutputProtocol, 0, (void**)&gop);
+    if (Status != EFI_SUCCESS)
+        ERR("UefiMachInit: Cannot setup GOP");
```
Comment (HBelusca) [inline boot/freeldr/uefi/uefisetup.c:45]: If you want to follow what was done also for XBOX and PC98 (see `XboxVideoInit` and `Pc98VideoInit`), this gop setup could be moved into the UefiInitializeVideo function. (If so you could make that latter one "VOID"; it wouldn't require the gop pointer as parameter.)

File: boot/freeldr/uefi/ueficon.c [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    else if (c == '\n')
+    {
+        CurrentCursorX = 0;
+
+        if (!NeedScroll)
+            ++CurrentCursorY;
+    }
+    else if (c == '\t')
+    {
+        CurrentCursorX = (CurrentCursorX + 8) & ~ 7;
```
Comment (HBelusca) [inline boot/freeldr/uefi/ueficon.c:51]:
```suggestion
        CurrentCursorX = (CurrentCursorX + 8) & ~7;
```

File: boot/freeldr/uefi/uefivid.c [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+
+UCHAR MachDefaultTextColor = COLOR_GRAY;
+REACTOS_INTERNAL_BGCONTEXT framebufferData;
+
+/* FUNCTIONS ******************************************************************/
+
+VOID
+UefiInitalizeVideo(_In_ EFI_GRAPHICS_OUTPUT_PROTOCOL* gop)
+{
+    gop->SetMode(gop,1);
```
[the remark this answers is not in the document]
Comment (DarkFire01) [reply boot/freeldr/uefi/uefivid.c:30]: It's just the lowest you can go.

GOP defines the video mode as an array you parse for the highest value.
We don't care about the higher video mode then the lowest fully supported by the firmware so we default to 1.

For example those devices that only have a single video mode will just use that mode, while my desktop has tons of modes including up to a 4K resolution.

I'll add a define for it so it isn't confusing.

File: boot/freeldr/CMakeLists.txt [hunk 15]
```diff
@@ -3,3 +3,4 @@ add_subdirectory(bootsect)
 add_subdirectory(fdebug)
 add_subdirectory(freeldr)
 add_subdirectory(install)
+add_subdirectory(uefi)
```
[the remark this answers is not in the document]
Comment (DarkFire01) [reply boot/freeldr/CMakeLists.txt:6]: I can do that, I wanted to separate it on this directory since its entirely separated from the compilation happening from within the freeldr directory.
e.g both uefi, freeldr, install all compile an object.
Whatever you prefer

File: boot/freeldr/uefi/uefivid.c [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+{
+    ULONG FgColor, BgColor;
+    UINT32 x, y;
+
+    UefiVideoAttrToColors(Attr, &FgColor, &BgColor);
+    for (y = 0; y < framebufferData.ScreenHeight; ++y)
+    {
+        for (x = 0; x < framebufferData.ScreenWidth; ++x)
+        {
+            *((UINT32*)(framebufferData.BaseAddress + 4 * framebufferData.PixelsPerScanLine * (y) + 4 * (x))) = BgColor;
```
Comment (HBelusca) [inline boot/freeldr/uefi/uefivid.c:90]: Is 4 == `sizeof(UINT32)` ? (If so, use sizeof instead of "4")

File: boot/freeldr/uefi/uefivid.c (hunk 5 above)
Comment (HBelusca) [reply boot/freeldr/uefi/uefivid.c:162]: See also comment above regarding the -2.

File: boot/freeldr/uefi/uefivid.c [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+UefiVideoSetDisplayMode(char *DisplayMode, BOOLEAN Init)
+{
+    /* We only have one mode, semi-text */
+    return VideoTextMode;
+}
+
+ULONG
+UefiVideoGetBufferSize(VOID)
+{
+    return ((framebufferData.ScreenHeight - 2) / CHAR_HEIGHT * (framebufferData.ScreenWidth / CHAR_WIDTH) * 2);
```
Comment (HBelusca) [inline boot/freeldr/uefi/uefivid.c:153]: Maybe "ScreenHeight - 2 * TOP_BOTTOM_LINES" (regarding the -2)? This is what was done in XBOX.
Also, UefiVideoScrollUp uses a similar construct.

File: boot/freeldr/freeldr/arch/uefi/uefisetup.c [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+    return Status;
+}
+
+VOID
+MachInit(const char *CmdLine)
+{
+    EFI_STATUS Status = 0;
+    Status = GlobalSystemTable->BootServices->AllocatePool(EfiLoaderData, sizeof(MachVtbl), (void**)&MachVtbl);
+    if (UefiHandleService(Status) != EFI_SUCCESS)
+        ERR("UefiMachInit: Could not allocated memory for MachVtbl");
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/uefi/uefisetup.c:38]: The MachVtbl is already an existing static variable, so that memory allocation should not be needed. What is actually going there is that you are storing in the first (4 or 8) bytes of the MachVtbl structure, a pointer value corresponding to that allocated memory, that you then just discard (when the table is zeroed out below) and leak.

File: boot/freeldr/freeldr/arch/uefi/uefisetup.c [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+
+    /* Setup GOP */
+    UefiInitalizeVideo();
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/uefi/uefisetup.c:41]: Move the gop allocation after all the table pointers are initialized, to follow what's done e.g. for xbox.

File: boot/freeldr/freeldr/arch/uefi/uefisetup.c [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+/* GLOBALS ********************************************************************/
+
+extern EFI_SYSTEM_TABLE* GlobalSystemTable;
+extern EFI_HANDLE GlobalImageHandle;
+BOOLEAN AcpiPresent = FALSE;
+
+/* FUNCTIONS ******************************************************************/
+
+EFI_STATUS
+UefiHandleService(EFI_STATUS Status)
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/uefi/uefisetup.c:22]: Is this function going to be used in the future?

File: boot/freeldr/freeldr/arch/uefi/uefildr.c [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+#include <uefildr.h>
+
+#include <debug.h>
+
+/* GLOBALS ********************************************************************/
+
+EFI_HANDLE GlobalImageHandle;
+EFI_SYSTEM_TABLE *GlobalSystemTable;
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/uefi/uefildr.c:15]:
```suggestion
EFI_SYSTEM_TABLE* GlobalSystemTable;
```

File: boot/freeldr/freeldr/arch/uefi/uefildr.c [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+
+/* FUNCTIONS ******************************************************************/
+
+EFI_STATUS
+EfiEntry(
+    _In_ EFI_HANDLE ImageHandle,
+    _In_ EFI_SYSTEM_TABLE *SystemTable)
+{
+    SystemTable->ConOut->OutputString(SystemTable->ConOut, L"UEFI EntryPoint: Starting freeldr from UEFI");
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/uefi/uefildr.c:24]: Could be removed in the future once UEFI boot works ^^

File: boot/freeldr/freeldr/arch/uefi/include/machuefi.h [hunk 23]
```diff
@@ -0,0 +1,119 @@
+/*
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/uefi/include/machuefi.h:1]: I'm thinking whether those two headers need to be moved into a `boot/freeldr/freeldr/include/arch/uefi` directory instead. Especially if there are going to be more in the future.
(see https://git.reactos.org/?p=reactos.git;a=tree;f=boot/freeldr/freeldr/include/arch;h=a0ea3edc72612364ec73886562a8c2487dab0333;hb=HEAD )

File: boot/freeldr/freeldr/arch/uefi/arch/amd64/misc.S [hunk 24]
```diff
@@ -0,0 +1,23 @@
+#include <asm.inc>
+#include <arch/pc/x86common.h>
+#include <arch/pc/pcbios.h>
+
+.code64
+
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/uefi/arch/amd64/misc.S:6]: Note: This helper file could be moved into the existing "boot/freeldr/freeldr/arch/amd64" directory instead, and you could move the definition for `__fastfail` from
https://git.reactos.org/?p=reactos.git;a=blob;f=boot/freeldr/freeldr/arch/amd64/entry.S;h=758eae4b677f3f3fb35c93a981c27cacb1eb26b0;hb=HEAD#l215
into it as well. (And in addition, remove those definitions from that entry.S file.)

File: boot/freeldr/freeldr/CMakeLists.txt [hunk 25]
```diff
@@ -1,4 +1,6 @@
 
+add_subdirectory(arch)
```
Comment (HBelusca) [inline boot/freeldr/freeldr/CMakeLists.txt:2]: Just do
```suggestion
add_subdirectory(arch/uefi)
```
if possible, that would remove the need of the lone cmakefiles in arch.

File: boot/freeldr/freeldr/arch/CMakeLists.txt [hunk 26]
```diff
@@ -0,0 +1 @@
+add_subdirectory(uefi)
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/CMakeLists.txt:1]: This is the "lone" cmakefiles ^^

File: boot/environ/CMakeLists.txt [hunk 27]
```diff
@@ -4,6 +4,7 @@ include_directories(BEFORE include/efi)
 include_directories(${REACTOS_SOURCE_DIR}/ntoskrnl/include)
 include_directories(${REACTOS_SOURCE_DIR}/sdk/lib/cmlib)
 include_directories(${REACTOS_SOURCE_DIR}/sdk/include/reactos/libs)
+include_directories(${REACTOS_SOURCE_DIR}/sdk/include/reactos/uefi)
```
Comment (HBelusca) [inline boot/environ/CMakeLists.txt:7]: I suppose that this change would better belong to a separate commit in a separate PR. I don't see yet any moving around of files or sthg related to uefi out of boot/environ.

File: boot/boot_images.cmake [hunk 28]
```diff
@@ -18,8 +18,8 @@ endif()
 # FIXME: this command creates a dummy EFI partition, add EFI/BOOT/boot${EFI_PLATFORM_ID}.efi file
 # once ReactOS supports UEFI
```
Comment (HBelusca) [inline boot/boot_images.cmake:19]: You can remove those two comment lines.

File: boot/boot_images.cmake [hunk 29]
```diff
@@ -162,4 +162,12 @@ add_custom_target(hybridcd
...
 
+# For things like flashing USB drives, we also add the efi file into efi/boot.
 add_cd_file(TARGET efisys FILE ${CMAKE_CURRENT_BINARY_DIR}/efisys.bin DESTINATION loader NO_CAB NOT_IN_HYBRIDCD FOR bootcd regtest livecd hybridcd)
+
+add_cd_file(
+    TARGET uefildr
+    DESTINATION efi/boot
+    NO_CAB
+    NAME_ON_CD boot${EFI_PLATFORM_ID}.efi
+    FOR livecd)
```
Comment (HBelusca) [inline boot/boot_images.cmake:173]: In principle it could be also for bootcd and hybridcd now.

File: boot/freeldr/freeldr/arch/uefi/CMakeLists.txt [hunk 30]
```diff
@@ -0,0 +1,210 @@
...
+    ../../ntldr/registry.c
+    ../../ntldr/winldr.c
+    ../../ntldr/wlmemory.c
+    ../../ntldr/wlregistry.c)
+
+list(APPEND FREELDR_ARC_SOURCE
+    ../../arcname.c
+    ../../arch/drivers/hwide.c
+    ../../arch/arcemul.c
+    ../../arch/archwsup.c
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/uefi/CMakeLists.txt:49]:
```suggestion
    ../drivers/hwide.c
    ../arcemul.c
    ../archwsup.c
```

File: boot/freeldr/freeldr/arch/uefi/CMakeLists.txt [hunk 31]
```diff
@@ -0,0 +1,210 @@
...
+list(APPEND FREELDR_BASE_ASM_SOURCE)
+
+if(ARCH STREQUAL "i386")
+    list(APPEND FREELDR_BOOTMGR_SOURCE
+        ../../arch/i386/i386bug.c)
+    list(APPEND FREELDR_BASE_ASM_SOURCE
+        ../../arch/i386/multiboot.S)
+
+    list(APPEND FREELDR_COMMON_ASM_SOURCE
+        ../../arch/i386/drvmap.S
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/uefi/CMakeLists.txt:87]: As for the other 3 files, those `../../arch/` can be replaced by just `../`.

Unless sometime in the near future you find the way how to share that list of files with the original freeldr's cmakelists, and avoid the copypaste. As done by e.g. ntoskrnl / ntkrnlmp.

File: boot/freeldr/freeldr/arch/uefi/uefisetup.c (hunk 20 above)
Comment (DarkFire01) [reply boot/freeldr/freeldr/arch/uefi/uefisetup.c:22]: Removed!

File: boot/freeldr/freeldr/arch/uefi/include/machuefi.h (hunk 23 above)
Comment (DarkFire01) [reply boot/freeldr/freeldr/arch/uefi/include/machuefi.h:1]: I've yet to add more headers even in the full loader branch, I'm going to do this anyway though I think it would be better.

File: boot/freeldr/freeldr/arch/uefi/uefivid.c [hunk 32]
```diff
@@ -0,0 +1,236 @@
...
+
+/* FUNCTIONS ******************************************************************/
+
+VOID
+UefiInitalizeVideo(VOID)
+{
+    EFI_STATUS Status;
+    EFI_GRAPHICS_OUTPUT_PROTOCOL* gop;
+
+    Status = GlobalSystemTable->BootServices->LocateProtocol(&EfiGraphicsOutputProtocol, 0, (void**)&gop);
```
[the remark this answers is not in the document]
Comment (DarkFire01) [reply boot/freeldr/freeldr/arch/uefi/uefivid.c]: GCC gets confused without it seem, expects to be of type (void**).

**Comment by HBelusca** [comment]
> I used reactos-gcc-i386-Debug-eb45fcb8a36d13619885c45f47d0fbb8ee79af31 (bootcd) downloaded here https://github.com/reactos/reactos/actions/runs/4361054372. It's on VirtualBox

Yes as you can see, freeldr now somewhat "starts" in UEFI mode. However, the PR is not the full UEFI implementation support for freeldr, but just part 1 over 10 or so...

File: boot/freeldr/freeldr/arch/amd64/misc.S [hunk 33]
```diff
@@ -0,0 +1,40 @@
+
+#include <asm.inc>
+#include <arch/pc/x86common.h>
+#include <arch/pc/pcbios.h>
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/amd64/misc.S:4]: I think those last two includes are unnecessary in this file.

File: boot/freeldr/freeldr/arch/uefi/uefisetup.c [hunk 34]
```diff
@@ -0,0 +1,58 @@
...
+extern EFI_SYSTEM_TABLE* GlobalSystemTable;
+extern EFI_HANDLE GlobalImageHandle;
+BOOLEAN AcpiPresent = FALSE;
+
+/* FUNCTIONS ******************************************************************/
+
+VOID
+MachInit(const char *CmdLine)
+{
+    MachVtbl.ConsPutChar = UefiConsPutChar;
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/uefi/uefisetup.c:24]:
```suggestion
    RtlZeroMemory(&MachVtbl, sizeof(MachVtbl));
    MachVtbl.ConsPutChar = UefiConsPutChar;
```

File: boot/freeldr/freeldr/arch/uefi/uefildr.c [hunk 35]
```diff
@@ -0,0 +1,36 @@
...
+EfiEntry(
+    _In_ EFI_HANDLE ImageHandle,
+    _In_ EFI_SYSTEM_TABLE *SystemTable)
+{
+    GlobalImageHandle = ImageHandle;
+    GlobalSystemTable = SystemTable;
+
+    BootMain(NULL);
+
+    //UNREACHABLE
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/uefi/uefildr.c:29]: Why commented out? This macro isn't available? (It should also have a semicolon after it.)

File: boot/freeldr/freeldr/arch/uefi/CMakeLists.txt [hunk 36]
```diff
@@ -0,0 +1,206 @@
+
+include_directories(BEFORE
```
Comment (HBelusca) [inline boot/freeldr/freeldr/arch/uefi/CMakeLists.txt:2]: I will open a separate PR just for the cmakelists thing.
